File: app/src/main/java/com/mxt/anitrend/base/custom/presenter/CommonPresenter.java
package com.mxt.anitrend.base.custom.presenter;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;

import com.mxt.anitrend.base.custom.recycler.RecyclerScrollListener;
import com.mxt.anitrend.base.interfaces.dao.BoxQuery;
import com.mxt.anitrend.base.interfaces.event.LifecycleListener;
import com.mxt.anitrend.data.DatabaseHelper;
import com.mxt.anitrend.extension.KoinExt;
import com.mxt.anitrend.util.Settings;

import org.greenrobot.eventbus.EventBus;

/**
 * Created by max on 2017/06/09.
 * Base presenter that will act as a template for all presenters
 * All preferences will be referenced from here.
 */

public abstract class CommonPresenter extends RecyclerScrollListener implements LifecycleListener {

    private Bundle bundle;
    private Context context;
    private BoxQuery databaseHelper;
    private Settings settings;
    
    public CommonPresenter(Context context) {
        this.context = context;
    }

    public Context getContext() {
        return context;
    }

    public Bundle getParams() {
        if(bundle == null)
            bundle = new Bundle();
        return bundle;
    }

    public void setParams(Bundle bundle) {
        this.bundle = bundle;
    }

    public BoxQuery getDatabase() {
        if(databaseHelper == null)
            databaseHelper = new DatabaseHelper();
        return databaseHelper;
    }

    /**
     * Unregister any listeners from fragments or activities
     */
    @Override
    public void onPause(SharedPreferences.OnSharedPreferenceChangeListener changeListener) {
        if(changeListener != null)
            getSettings().getSharedPreferences()
                    .unregisterOnSharedPreferenceChangeListener(changeListener);
    }

    /**
     * Register any listeners from fragments or activities
     */
    @Override
    public void onResume(SharedPreferences.OnSharedPreferenceChangeListener changeListener) {
        if(changeListener != null)
            getSettings().getSharedPreferences()
                    .registerOnSharedPreferenceChangeListener(changeListener);
    }

    /**
     * Destroy any reference which maybe attached to
     * our context
     */
    @Override
    public void onDestroy() {
        bundle = null;
    }

    public Settings getSettings() {
        if(settings == null)
            settings = KoinExt.get(Settings.class);
        return settings;
    }

    /**
     * Trigger all subscribers that may be listening. This method makes use of sticky broadcasts
     * in case all subscribed listeners were not loaded in time for the broadcast
     * <br/>
     *
     * @param param the object of type T to send
     * @param sticky set true to make sticky post
     */
    public <T> void notifyAllListeners(T param, boolean sticky) {
        if(sticky)
            EventBus.getDefault().postSticky(param);
        else
            EventBus.getDefault().post(param);
    }

    public interface AbstractPresenter <S extends CommonPresenter> {
        S getPresenter();
    }
}

File: app/src/main/java/com/mxt/anitrend/base/custom/presenter/CommonPresenter.kt
package com.mxt.anitrend.base.custom.presenter

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import com.mxt.anitrend.base.custom.recycler.RecyclerScrollListener
import com.mxt.anitrend.base.interfaces.dao.BoxQuery
import com.mxt.anitrend.base.interfaces.event.LifecycleListener
import com.mxt.anitrend.data.DatabaseHelper
import com.mxt.anitrend.extension.KoinExt
import com.mxt.anitrend.util.Settings

/**
 * Created by max on 2017/06/09.
 * Base presenter that will act as a template for all presenters
 * All preferences will be referenced from here.
 */
abstract class CommonPresenter(
    val context: Context,
    val database: BoxQuery,
    val settings: Settings,
) : RecyclerScrollListener(),
    LifecycleListener {
    private var bundle: Bundle? = null

    var params: Bundle
        get() {
            if (bundle == null) {
                bundle = Bundle()
            }
            return requireNotNull(bundle)
        }
        set(value) {
            bundle = value
        }

    /**
     * Unregister any listeners from fragments or activities
     */
    override fun onPause(changeListener: SharedPreferences.OnSharedPreferenceChangeListener?) {
        changeListener?.let { settings.unregisterOnSharedPreferenceChangeListener(it) }
    }

    /**
     * Register any listeners from fragments or activities
     */
    override fun onResume(changeListener: SharedPreferences.OnSharedPreferenceChangeListener?) {
        changeListener?.let { settings.registerOnSharedPreferenceChangeListener(it) }
    }

    /**
     * Destroy any reference which maybe attached to
     * our context
     */
    override fun onDestroy() {
        bundle = null
    }

    interface AbstractPresenter<S : CommonPresenter> {
        fun getPresenter(): S
    }
}
